### PR TITLE
Add cloaked CNAME detection

### PR DESCRIPTION
**Pull request by joshliebe**
Task/Issue URL: https://app.asana.com/0/0/1202737208022388/f

### Description
This PR uses the CNAMEs section in the block list to block trackers that are using CNAME cloaking via Content Blocking rules.

### Steps to test this PR

- [x] Visit duluthtrading.com
- [x] Verify `duluthtrading.com.ssl.d2.sc.omtrdc.net` is listed as a blocked tracker (Cloaked host: metrics.duluthtrading.com)
- [x] Visit sky.com
- [x] Verify `sky.com.ssl.d1.sc.omtrdc.net` is listed as a blocked tracker (Cloaked host: smetrics.sky.com)
- [x] Visit other well known websites
- [x] Verify standard functionality is not affected

### CODE REVIEW

File: app/src/androidTest/resources/reference_tests/domain_matching_tests.json [hunk 1]
```diff
@@ -77,10 +77,7 @@
         "siteURL": "https://randomsite123.com/",
         "requestURL": "https://bad.cnames.test/something",
         "requestType": "script",
-        "expectAction": "block",
-        "exceptPlatforms": [
-          "android-browser"
-        ]
+        "expectAction": "block"
```
Comment (joshliebe) [inline app/src/androidTest/resources/reference_tests/domain_matching_tests.json:80]: PR for this change here: https://github.com/duckduckgo/privacy-reference-tests/pull/56

File: app/src/main/java/com/duckduckgo/app/trackerdetection/CloakedCnameDetector.kt [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    fun detectCnameCloakedHost(url: Uri): String?
+}
+
+@ContributesBinding(AppScope::class)
+@SingleInstanceIn(AppScope::class)
+class CloakedCnameDetectorImpl @Inject constructor(
+    private val tdsCnameEntityDao: TdsCnameEntityDao
+) : CloakedCnameDetector {
+
+    override fun detectCnameCloakedHost(url: Uri): String? {
```
Comment (marcosholgado) [inline app/src/main/java/com/duckduckgo/app/trackerdetection/CloakedCnameDetector.kt:37]: Should we test this? :)

File: app/src/main/java/com/duckduckgo/app/trackerdetection/TrackerDataLoader.kt [hunk 3]
```diff
@@ -88,6 +90,7 @@ class TrackerDataLoader @Inject constructor(
             tdsEntityDao.updateAll(tdsJson.jsonToEntities())
             tdsDomainEntityDao.updateAll(tdsJson.jsonToDomainEntities())
             tdsTrackerDao.updateAll(tdsJson.jsonToTrackers().values)
+            tdsCnameEntityDao.updateAll(tdsJson.jsonToCnameEntities())
```
Comment (marcosholgado) [inline app/src/main/java/com/duckduckgo/app/trackerdetection/TrackerDataLoader.kt:93]: Should we test that we are now doing this? :)

File: app/src/main/java/com/duckduckgo/app/trackerdetection/db/TdsCnameEntityDao.kt [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    abstract fun getAll(): List<TdsCnameEntity>
+
+    @Query("select * from tds_cname_entity where cloakedHostName=:cloakedHostName")
+    abstract fun get(cloakedHostName: String): TdsCnameEntity?
+
+    @Query("delete from tds_cname_entity")
+    abstract fun deleteAll()
+
+    @Query("select count(*) from tds_cname_entity")
+    abstract fun count(): Int
```
Comment (marcosholgado) [inline app/src/main/java/com/duckduckgo/app/trackerdetection/db/TdsCnameEntityDao.kt:38]: I bet you don't need this one :D

File: app/src/main/java/com/duckduckgo/app/browser/WebViewRequestInterceptor.kt [hunk 5]
```diff
@@ -150,21 +152,39 @@ class WebViewRequestInterceptor(
...
         val trackingEvent = trackingEvent(request, documentUrl, webViewClientListener)
         if (trackingEvent?.status == TrackerStatus.BLOCKED) {
-            trackingEvent.surrogateId?.let { surrogateId ->
-                val surrogate = resourceSurrogates.get(surrogateId)
-                if (surrogate.responseAvailable) {
-                    Timber.d("Surrogate found for ${request.url}")
-                    webViewClientListener?.surrogateDetected(surrogate)
-                    return WebResourceResponse(surrogate.mimeType, "UTF-8", surrogate.jsFunction.byteInputStream())
+            return blockRequest(trackingEvent, request, webViewClientListener)
+        } else if (trackingEvent == null ||
```
Comment (marcosholgado) [inline app/src/main/java/com/duckduckgo/app/browser/WebViewRequestInterceptor.kt:156]: Should we test this block?

**Review by marcosholgado — APPROVED** [review]
Left a few comments, mainly about testing. Otherwise LGTM, nice job!

File: app/src/main/java/com/duckduckgo/app/trackerdetection/TrackerDataLoader.kt (hunk 3 above)
Comment (joshliebe) [reply app/src/main/java/com/duckduckgo/app/trackerdetection/TrackerDataLoader.kt:93]: I've added a test for this function (Class was untested).

File: app/src/main/java/com/duckduckgo/app/browser/WebViewRequestInterceptor.kt (hunk 5 above)
Comment (joshliebe) [reply app/src/main/java/com/duckduckgo/app/browser/WebViewRequestInterceptor.kt:156]: Added tests for this 👍

File: app/src/main/java/com/duckduckgo/app/trackerdetection/db/TdsCnameEntityDao.kt (hunk 4 above)
Comment (joshliebe) [reply app/src/main/java/com/duckduckgo/app/trackerdetection/db/TdsCnameEntityDao.kt:38]: I'm going to keep it though (It's used in the test 😁)

**Pull request merged**